File: lista4/zad3-rust/src/lib/filozof.rs
use rand::{rngs::ThreadRng, Rng};
use std::sync::{Arc, Mutex, MutexGuard};
pub struct Filozof {
    pub id: usize,
    pub lewy: Arc<Mutex<()>>,
    pub prawy: Arc<Mutex<()>>,
    pub do_zjedzenia: usize,
}

impl Filozof {
    pub fn new(
        id: usize,
        lewy: Arc<Mutex<()>>,
        prawy: Arc<Mutex<()>>,
        do_zjedzenia: usize,
    ) -> Self {
        Self {
            id,
            lewy,
            prawy,
            do_zjedzenia,
        }
    }

    fn wez_lewy(&self) {
        //-> MutexGuard<()> {
        println!("{}[{}] bierze lewy widelec", " ".repeat(self.id), self.id);
        //self.lewy.lock().unwrap()
    }

    fn wez_prawy(&self) {
        //-> MutexGuard<()> {
        println!("{}[{}] bierze prawy widelec", " ".repeat(self.id), self.id);
        //self.prawy.lock().unwrap()
    }

    fn jedz(&self, rng: &mut rand::rngs::ThreadRng) {
        println!(
            "{}[{}] je (zostało {} obiadów)",
            " ".repeat(self.id),
            self.id,
            self.do_zjedzenia - 1
        );
        std::thread::sleep(std::time::Duration::from_millis(rng.gen_range(0..1000)));
    }

    fn najedzony(&self) -> bool {
        self.do_zjedzenia == 0
    }

    fn mysl(&self, rng: &mut rand::rngs::ThreadRng) {
        println!("{}[{}] mysli", " ".repeat(self.id), self.id);
        std::thread::sleep(std::time::Duration::from_millis(rng.gen_range(0..1000)));
    }

    pub fn zjedz_wszystko(&mut self, mut rng: ThreadRng) {
        while !self.najedzony() {
            self.mysl(&mut rng);

            let _lewy = self.lewy.lock().unwrap();
            self.wez_lewy();

            let _prawy = self.prawy.lock().unwrap();
            self.wez_prawy();

            self.jedz(&mut rng);
            self.do_zjedzenia -= 1;

            drop(_lewy);
            println!("{}[{}] odkłada lewy widelec", " ".repeat(self.id), self.id);

            drop(_prawy);
            println!("{}[{}] odkłada prawy widelec", " ".repeat(self.id), self.id);
        }
    }
}

File: lista4/zad2-go/filozof.go
package main

import (
	"fmt"
	"math/rand"
	"strings"
	"sync"
	"time"
)

type filozof struct {
	id          int
	lewy, prawy *widelec
	doZjedzenia int
}

type kolej chan struct{}

func newFilozof(id int, lewy, prawy *widelec) filozof {
	return filozof{id, lewy, prawy, doZjedzenia}
}

func (f *filozof) wezLewy() {
	f.lewy.Lock()
}

func (f *filozof) odlozLewy() {
	f.lewy.Unlock()
}

func (f *filozof) wezPrawy() {
	f.prawy.Lock()
}

func (f *filozof) odlozPrawy() {
	f.prawy.Unlock()
}

func (f *filozof) jedz() {
	f.doZjedzenia--
}

func (f *filozof) najedzony() bool {
	return f.doZjedzenia == 0
}

func (f *filozof) jedzenie(wg *sync.WaitGroup, kolej *kolejka) {
	defer wg.Done()

	spaces := strings.Repeat(" ", f.id)

	for !f.najedzony() {
		kolej.wez()

		fmt.Printf("%s[%d] mysli\n", spaces, f.id)
		f.mysli()
		f.wezLewy()
		fmt.Printf("%s[%d] bierze lewy widelec\n", spaces, f.id)
		f.wezPrawy()
		fmt.Printf("%s[%d] bierze prawy widelec\n", spaces, f.id)
		fmt.Printf("%s[%d] je (zostało %d obiadów)\n", spaces, f.id, f.doZjedzenia-1)
		f.mysli()
		f.jedz()
		f.odlozLewy()
		fmt.Printf("%s[%d] odklada lewy widelec\n", spaces, f.id)
		f.odlozPrawy()
		fmt.Printf("%s[%d] odklada prawy widelec\n", spaces, f.id)

		kolej.odloz()
	}
}

func (f *filozof) mysli() {
	t := time.Duration(rand.Int63n(randSecond))
	time.Sleep(t)
}
